src/AddThreepid.js

```diff
@@ -171,12 +174,31 @@ export default class AddThreepid {
                         id_access_token: identityAccessToken,
                     });
                 } else {
-                    await MatrixClientPeg.get().addThreePidOnly({
-                        sid: this.sessionId,
-                        client_secret: this.clientSecret,
-                    });
+                    try {
+                        await this._makeAddThreepidOnlyRequest();
+
+                        // The spec has always required this to use UI auth but synapse briefly
+                        // implemented it without, so this may just succeed and that's OK.
+                        return;
+                    } catch (e) {
+                        if (e.httpStatus !== 401 || !e.data || !e.data.flows) {
+                            // doesn't look like an interactive-auth failure
+                            throw e;
+                        }
+
+                        // pop up an interactive auth dialog
+                        const InteractiveAuthDialog = sdk.getComponent("dialogs.InteractiveAuthDialog");
+
+                        Modal.createTrackedDialog('Add Email', '', InteractiveAuthDialog, {
```

Comment:
Callers of this function are expecting a promise that resolves when the add request completes. I _think_ returning the modal's finished promise is enough to wait on the request here?

Reply:
Good point - done.